consult-yasnippet.el

```diff
@@ -0,0 +1,140 @@
+;; GNU General Public License for more details.
+
+;; You should have received a copy of the GNU General Public License
+;; along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+;;; Commentary:
+
+;; Provides the command `consult-yasnippet' which presents all the snippet
+;; expansions available in the current buffer with in-buffer previews.
+;;
+;; This is an extra package, to avoid loading yasnippet in the base consult
+;; package..
+
+
+;;; Code:
+
+(require 'consult)
+(require 'yasnippet)
+
+(defun consult-yasnippet--expand-template (template region region-contents)
+  "Expand the yasnippet template TEMPLATE at point."
+  (deactivate-mark)
+  (goto-char (car region))
+  (when (not (string-equal "" region-contents))
+    (push-mark (point))
+    (push-mark (cdr region) nil t))
+
+  (yas-expand-snippet (yas--template-content template)
+                      nil nil
+                      (yas--template-expand-env template)))
+
+(defun consult-yasnippet--preview ()
+  "Previewer for `consult--read'.
+This function expands TEMPLATE at point in the buffer
+`consult-yasnippet--read-snippet' was started in. This includes
+overwriting any region that was active and removing any previous
+previews that're already active.
+
+When TEMPLATE is not given, this function essentially just resets
+the state of the current buffer to before any snippets were previewed."
+  (let* ((buf (current-buffer))
+         (region (if (region-active-p)
+                     (cons (region-beginning) (region-end))
+                   (cons (point) (point))))
+         (region-contents (buffer-substring (car region) (cdr region))))
+    (lambda (template restore)
+      (with-current-buffer buf
+        (let ((yas-verbosity 0)
+              (inhibit-redisplay t)
+              (inhibit-read-only t)
+              (orig-offset (- (point-max) (cdr region)))
+              (yas-prompt-functions '(yas-no-prompt)))
+
+          ;; We always undo any snippet previews before maybe setting up
+          ;; some new previews.
+          (delete-region (car region) (cdr region))
+          (goto-char (car region))
+          (setcar region (point))
+          (insert region-contents)
+          (setcdr region (point))
+
+          (when (and template (not restore))
+            (unwind-protect
+                (consult-yasnippet--expand-template template region region-contents)
+              (unwind-protect
+                  (mapc #'yas--commit-snippet
+                        (yas-active-snippets (point-min) (point-max)))
+                (setcdr region (- (point-max) orig-offset))))
+            (redisplay)))))))
+
+(defun consult-yasnippet--candidates (&rest body)
+  "Retrieve the list of available snippets in the current buffer."
+  (unless (bound-and-true-p yas-minor-mode)
+    (error "`yas-minor-mode' not enabled in current buffer"))
+
+  (cl-loop for template in (yas--all-templates (yas--get-snippet-tables))
```

Comment:
Please rewrite this with mapcar. I am currently not using cl-loop in consult. It is only a stylistic thing.

Reply:
I try to avoid using mapcar whenever I end up using a lambda with it because of performance concerns (I seem to recall reading somewhere that lambda evaluations due to closures slow down evaluation). But if its a style thing then I hope you don't mind I've changed it to a dolist,push,reverse operation.

Reply:
Yes it slows down, but it is marginal in this case. You can safely use a mapcar here.

Reply:
Is what I've currently submitted still within the style guidelines or would you prefer I update it to use `mapcar`?

Reply:
I don't really like it since it is just a clouded mapcar. I also have such code in consult.el but only where performance matters. However these things are minor nitpicks, I can also change it myself one way or the other if I feel like it. More important for me is to understand what is going on and this has not yet been addressed sufficiently. I will add some more comments to the current version to mark the places which I find hard to follow. I understand that you copied it from ivy-yasnippet, but you could also ask the original author if they are interested in a port to consult

Reply:
All right, switched to `mapcar` to save you the hassle of changing it sometime down the line.

I'll open an issue on `ivy-yasnippet` and ask whether they'd be willing to help out :-).